programs/bpf_loader/src/syscalls.rs

```diff
@@ -2663,26 +2664,51 @@ where
                 invoke_context,
             )?;
             {
-                callee_account
-                    .set_lamports(*caller_account.lamports)
-                    .map_err(SyscallError::InstructionError)?;
-                callee_account
-                    .set_data(caller_account.data)
-                    .map_err(SyscallError::InstructionError)?;
-                callee_account
-                    .set_executable(caller_account.executable)
-                    .map_err(SyscallError::InstructionError)?;
-                callee_account
-                    .set_owner(caller_account.owner.as_ref())
-                    .map_err(SyscallError::InstructionError)?;
+                if callee_account.get_lamports() != *caller_account.lamports {
+                    callee_account
+                        .set_lamports(*caller_account.lamports)
+                        .map_err(SyscallError::InstructionError)?;
+                }
+                // The redundant `can_data_be_resized` helps to avoid the expensive data comparison if we can
+                if callee_account
+                    .can_data_be_resized(caller_account.data.len())
+                    .is_ok()
+                    || callee_account.get_data() != caller_account.data
+                {
```

Comment:
This is a complex one for sure, and it was carefully crafted to minimize the performance overhead (before and after feature activation) and not require a feature gate (stay invariant to the feature activation). For a detailed explaination see: https://github.com/solana-labs/solana/pull/25899#discussion_r897409117.

> the name of the `can_data_be_resized` method doesn't fully indicate that `set_data` will succeed.

The comment for `can_data_be_resized` states that, but I agree that the naming could be better. Suggestions?

> We should skip the data comparison if the data length changed in an invalid way

Is already the case, if the data length changed in an invalid way `can_data_be_resized` errors and the data comparison never happens.

> We should reuse the errors from check a) if the data was changed rather than recomputing that error inside set_data.

In ABIv1, because of the workaround which was necessary to make it work with `PreAccount`, all accounts are always written back. So readonly accounts produce incorrect errors and we need to ignore not (re)use them.

Reply:
> The comment for can_data_be_resized states that, but I agree that the naming could be better. Suggestions?

I think your names are fine, it's just that I don't like that `can_data_be_resized` includes the behavior of `can_data_be_changed` as well. I think that the redundant check should check if both `can_data_be_changed` and `can_data_be_resized` return ok before skipping the data comparison.

> Is already the case, if the data length changed in an invalid way can_data_be_resized errors and the data comparison never happens.

When `can_data_be_resized` errors, we always continue to line 2676 which does a data comparison

> In ABIv1, because of the workaround which was necessary to make it work with PreAccount, all accounts are always written back. So readonly accounts produce incorrect errors and we need to ignore not (re)use them.

I understand that. I'm talking about something slightly different. In case 2 above, we know that the data change will fail if the data comparison returns not-equal, so we should reuse that error that we already computed by calling the "redundant" `can_data_be_resized` method.

Reply:
I split `can_data_be_resized` and `can_data_be_changed` to be independent of one another. And I moved `can_data_be_resized()?` out of the `if` so that it can lead to early termination.

Reply:
The new code only detects invalid resizes for unowned accounts. If an owned executable or read-only account is resized, we will do a full data comparison even though checking the length is sufficient. Might be good to fix that by adding more checks in `can_data_be_resized`?

And for the record, here's the code that I think would be more readable:
```rust
        // The redundant check helps to avoid the expensive data comparison if we can
        match borrowed_account.can_data_be_changed() {
            Ok(()) => borrowed_account.set_data(&data)?,
            Err(err) => {
                let current_data = borrowed_account.get_data();
                if current_data.len() != data.len() || current_data != data {
                    return Err(err);
                }
            }
        }
```

Reply:
@Lichtso I see you resolved this, anything to comment about the edge cases I mentioned above for avoiding data comparison?

Reply:
Added your `match` based code. But `can_data_be_resized()` needs to stay and come before `can_data_be_changed()` otherwise the error priority of `InstructionError::AccountDataSizeChanged` would change.